[Usage-based] Update billing page layout for teams and personal accounts

jankeromnes opened the pull request:
## Description
<!-- Describe your changes in detail -->

Update billing page layout for usage-based teams and personal accounts.

## Related Issue(s)
<!-- List the issue(s) this PR solves -->
Fixes https://github.com/gitpod-io/gitpod/issues/13405

## How to test
<!-- Provide steps to test this PR -->

1. Create a team called "Gitpod [Something]"
2. Check the Team Billing UI
3. Upgrade the team by adding a credit card
4. Check the Team Billing UI again
5. Go to `/billing` (your individual billing settings)
6. Check your User Billing UI
7. Upgrade your user by adding a credit card
8. Check your User Billing UI again

## Release Notes
<!--
  Add entries for the CHANGELOG.md or "NONE" if there aren't any user facing changes.
  Each line becomes a separate entry.
  Format: [!<optional for breaking>] <description>
  Example: !basic auth is no longer supported
  See https://www.notion.so/gitpod/Release-Notes-513a74fdd23b4cb1b3b3aefb1d34a3e0
-->
```release-note
NONE
```

## Documentation
<!--
Does this PR require updates to the documentation at www.gitpod.io/docs?
* Yes
  * 1. Please create a docs issue: https://github.com/gitpod-io/website/issues/new?labels=documentation&template=DOCS-NEW-FEATURE.yml&title=%5BDocs+-+New+Feature%5D%3A+%3Cyour+feature+name+here%3E
  * 2. Paste the link to the docs issue below this comment
* No
  * Are you sure? If so, nothing to do here.
-->

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [x] /werft with-payment
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

jankeromnes commented:
/werft run

jankeromnes commented:
/werft run

jankeromnes commented:
## Screenshots

| Billing Mode | Light | Dark |
| --- | --- | --- |
| Team Billing (before upgrade) | <img width="1437" alt="Screenshot 2022-10-06 at 10 47 12" src="https://user-images.githubusercontent.com/599268/194268270-87531364-81b3-4514-8e68-9e2f954c1a36.png"> | <img width="1434" alt="Screenshot 2022-10-06 at 10 47 30" src="https://user-images.githubusercontent.com/599268/194268281-0e257b8c-eecb-4fd7-8640-70cdb4f028fb.png"> |
| Team Billing (after upgrade) | <img width="1436" alt="Screenshot 2022-10-06 at 10 49 17" src="https://user-images.githubusercontent.com/599268/194268282-b7296b24-a9af-4a22-8959-7c3ddb2c47e8.png"> | <img width="1437" alt="Screenshot 2022-10-06 at 10 49 49" src="https://user-images.githubusercontent.com/599268/194268287-3cd3e791-ee32-4d80-a479-08d1daaa689a.png"> |
| User Billing (before upgrade) | <img width="1436" alt="Screenshot 2022-10-06 at 10 50 26" src="https://user-images.githubusercontent.com/599268/194268293-a6a1ef8e-dfb1-4050-8ea7-3c4ec5952729.png"> | <img width="1436" alt="Screenshot 2022-10-06 at 10 50 47" src="https://user-images.githubusercontent.com/599268/194268297-ae29fa94-d6bd-41a4-9a47-8eba763e64a5.png"> |
| User Billing (after upgrade) | <img width="1434" alt="Screenshot 2022-10-06 at 10 51 57" src="https://user-images.githubusercontent.com/599268/194268763-f23cd8a3-77f2-4f1c-b372-476ee12c0d11.png"> | <img width="1432" alt="Screenshot 2022-10-06 at 10 52 13" src="https://user-images.githubusercontent.com/599268/194268778-a9a56d00-069f-41d2-a3fb-7142efa0e86f.png"> |

laushinka commented:
Thanks, @jankeromnes! 
For some time this was `NaN` when I upgraded my team.
When I came back from another page it was correctly a number.

<img width="228" alt="Screenshot 2022-10-06 at 12 04 49" src="https://user-images.githubusercontent.com/8015191/194286378-9230c52c-fd88-437e-adf3-82729adb1878.png">

jankeromnes commented:
>  For some time this was `NaN` when I upgraded my team.

Aha, good catch @laushinka! This happens when the usage limit is `0`. To be honest, I'm not quite sure what we're supposed to show here when the limit is `0`... 🤔

But the `NaN` definitely looks weird. Maybe let's hide the percentage when it doesn't make sense. 👍

laushinka commented:
> This happens when the usage limit is `0` (divide-by-zero gives NaN). 

How about when the limit is 0, instead of "Manage usage limit" we show "Set usage limit" and hide the percentage?

laushinka commented:
> But doesn't 0 currently mean "I want absolutely no usage"? (I.e. when it's set to 0, I believe Gitpod won't start any workspaces, and you'll see the "limit reached" notification.)
> 
> Personally, I'd be happy with revisiting the semantics of `usageLimit = 0`, e.g. redefine it as "unlimited usage"?

Ahh, right. No limit means unlimited usage, but limit zero means no usage. Then let's go with your initial suggestion 👍🏼

gtsiolis reviewed (approved):
Excited to see this feature coming together! ✨ 

Thanks, @jankeromnes!

Left some comments below, hopefully not too overwhelming. 🏁 

Let's open follow-up issues for anything that feels out of the scope of this PR and could significantly increase the scope of the changes here. 🏓

jankeromnes commented:
Woohoo, many thanks @gtsiolis for the extensive UX review!

/hold so that I get a chance to address your comments and suggestions 👀

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                        </div>
+                    </div>
+                )}
+                {showUpgradeTeam && (
```
Good catch! But I think it's not easy to do away with them -- the first one is when we figure out whether to show the old Chargebee UI or the new Stripe UI, and the second one is while we actually load all the Stripe-relevant data.

jankeromnes replied:
Or, we could simply decide not to show the initial spinner. I generally don't like leaving any "loading states" as a blank page ("wait, that's it? is it doing anything? is it broken?") but if it always transitions rapidly to the Stripe Billing UI, then maybe we could do without the first spinner. 💭

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                                    50 hours of Standard workspace usage.{" "}
+                                    <a
+                                        className="gp-link"
+                                        href="https://www.gitpod.io/docs/configure/billing/usage-based-billing"
```
Thanks! I initially had that, but then I noticed the entire page has like one intro paragraph before that section, so we might as well show users the intro, too. 😊 But happy to adjust if needed.

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                            {AttributionId.parse(attributionId)?.kind === "user" ? (
+                                <>
+                                    <div className="mt-1 text-xl font-semibold flex-grow text-gray-800 dark:text-gray-100">
+                                        {currency === "EUR" ? "€" : "$"}9 / month
```
In preview environments, everyone always sees $ as the default. In production, the currency is pre-selected based on your region. This auto-select is just not configured for preview environments or staging.

gtsiolis replied:
Ah, I see. Thanks for the insight! 💡 

Do you think this is trivial or worth the effort to replicate this in preview environments?

jankeromnes replied:
Definitely not trivial, so probably not worth the effort. A cheap fix could be to make preview always show € instead of $ 🤷 but maybe that's also not really worth the effort. Please feel free to open an issue if you'd like this to be fixed. 😇

jankeromnes commented on components/dashboard/src/settings/Billing.tsx:
```diff
@@ -13,15 +13,16 @@ import { AttributionId } from "@gitpod/gitpod-protocol/lib/attribution";
-    const attributionId: AttributionId = { kind: "user", userId: user?.id || "" };
 
     return (
         <PageWithSettingsSubMenu title="Billing" subtitle="Usage-Based Billing.">
```
Note: This is the title shown at the top of the page. Also, I left the section headings in the body of the page for now, but I'm happy to remove them (I wasn't quite sure if they were still needed, because the specs only focus on the contents, and not the general context around them -- so I didn't see if there was still a title above it or not)

gtsiolis replied:
> I wasn't quite sure if they were still needed, because the specs only focus on the contents, and not the general context around them

Ah, ok! I thought it was clear that these specs included only the parts we're changing (see [relevant discussion](https://gitpod.slack.com/archives/C01KPEPQYE6/p1664795185245119?thread_ts=1664363685.869519&cid=C01KPEPQYE6) (internal)), but you've got a valid point, there was no decision documented around dropping the section headers.

Let's keep them for now and see if someone complains about this change, we can always drop them later, which would take less than 5 minutes to make, see [relevant discussion](https://gitpod.slack.com/archives/C01KPEPQYE6/p1664538609541969?thread_ts=1664363685.869519&cid=C01KPEPQYE6) (internal). 🤝

jankeromnes replied:
Thanks! I'm more than happy to delete the titles. However, I get lost in all our discussions, meeting notes, recordings, Notion pages, and PR comments 😅 

Could you please take a few quick screenshots, and obviously highlight what needs to be removed? Then it will be really easy for me to do so.

gtsiolis replied:
No need to change anything for now. Let's rather keep the section headers as they are, as removing the section headers requires us to introduce a better dropdown element for the billing account. Dropping the headers now would be quite confusing. 🤝

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                        <div className="mt-2 flex">
+                            <progress className="h-4 flex-grow rounded-xl" value={currentUsage} max={usageLimit} />
+                        </div>
+                        <div className="bg-gray-100 dark:bg-gray-800 border-t border-gray-200 dark:border-gray-800 -m-4 p-4 mt-4 rounded-b-xl flex">
```
Thanks for suggesting better colors! However, in order to make implementation easy, I would love us to stick to a consistent mapping between light and dark colors.

Currently in our code base,
- with `bg-gray-100`, we use mostly `dark:bg-gray-800`, but occasionally also `dark:bg-gray-700` and `dark:bg-gray-900` -- it would be nice to make all of these consistent if possible
- with `border-gray-200`, we use overwhelmingly `dark:border-gray-800`, but also `dark:border-gray-700` and `dark:border-gray-600` once or twice

| With `bg-gray-100` | With `border-gray-200` |
| --- | --- |
| <img width="559" alt="Screenshot 2022-10-07 at 11 00 56" src="https://user-images.githubusercontent.com/599268/194516657-63c074a3-67d6-46cf-9018-790690d61473.png"> | <img width="563" alt="Screenshot 2022-10-07 at 11 03 48" src="https://user-images.githubusercontent.com/599268/194516671-f14242ce-807e-47d8-808c-c00b54de280d.png"> |

Maybe this can be a follow-up issue, but I would be very happy to improve our consistency here. 😌

jankeromnes replied:
Follow-up thought: It does occur to me that, when we want one shade, we'll typically go for `bg-gray-100 dark:bg-gray-800`, but if we need two shades (e.g. "combined box with separator") maybe we want the second shade to be lighter both in light theme and dark theme?

I.e. we'll add the second shade `bg-gray-50`, but not `dark:bg-gray-900` -- instead, we'd shift the existing `dark:bg-gray-800` to `dark:bg-gray-700` and re-use `dark:bg-gray-800` as the "second shade"?

This would lead to inconsistencies between:
- `bg-gray-100 dark:bg-gray-800` alone (one background shade)
- `bg-gray-50 dark:bg-gray-800` combined with `bg-gray-100 dark:bg-gray-700`

Long story short, maybe we simply cannot have a 100% consistent mapping between light and dark shades, because we're using them differently in different contexts? 🤔

gtsiolis replied:
I understand the concerns, especially since from a technical perspective this may look inconsistent or repetitive, but the colors used depend also on the layer they are used and the maturity of each component as they may evolve over time. This will get better as we create more reusable components and the product and components evolves. ➿ 

Like you [mentioned above](https://github.com/gitpod-io/gitpod/pull/13564#discussion_r989931274), because they are used in different contexts and sometimes different layers, it's impossible to have one mapping to rule them all to deliver a good UX. We need more grays!

Last but not least, the [screenshots above](https://github.com/gitpod-io/gitpod/pull/13564#discussion_r989866848) are a great example of :a: how we practice our values[[1](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#c2cdfc9a443644dd99e87031c375fbf3)][[2](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#30d4202757c74abc9821a1c79e0b1f75)][[3](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#2e0e717c03c04de882012e60ca348adc)][[4](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#a6c78fe159d547deb9426f30b7017aef)][[5](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#c8f962af695c4a43b4f39bdd3f5363a6)][[6](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#d766b4a2f4664dea9905ac29e65a5ccf)][[7](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#0c2e0d6c8a99404b97ccdc206b573d72)][[8](https://www.notion.so/gitpod/Core-Values-Differentiators-2ed4c2f93c84499b98e3b5389980992e#5c98845e99264744bf9c70ad710a1397)] and :b: what enabled us to make all the progress we've made so far in this small amount of time.

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                        <div className="bg-gray-100 dark:bg-gray-800 border-t border-gray-200 dark:border-gray-800 -m-4 p-4 mt-4 rounded-b-xl flex">
+                            <div className="flex-grow">
+                                <div className="uppercase text-sm text-gray-400 dark:text-gray-500">Current Period</div>
+                                <div className="text-sm font-medium text-gray-500 dark:text-gray-400">
+                                    {now.toLocaleString("default", { month: "long" })} {now.getFullYear()}
+                                </div>
```
It's a little bit more code, but shouldn't be a problem. 😊

gtsiolis replied:
If it takes more than a few minutes, let's open a follow-up issue for this! 🏓

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                            <Check className="m-0.5 w-5 h-5" />
+                            <div className="flex flex-col">
+                                <span>
+                                    36¢ for 10 credits or 1 hour of Standard workspace usage.{" "}
```
The official names of the configured classes in production are actually "Standard" and "Large" ([internal source](https://github.com/gitpod-io/ops/blob/208e9f7b2dc264c68a01a89ef763de43f680315e/deploy/production/meta-eu02/app/installer-config.yaml#L186-L202)):

<img width="750" alt="Screenshot 2022-10-07 at 12 53 58" src="https://user-images.githubusercontent.com/599268/194537498-4821f14d-cb3a-49b0-a7a0-9fcc4c2eca42.png">

Maybe it's just the preview environments that are configured differently? If that's the case, I think we can safely ignore this inconsistency, or maybe fix it by aligning the configured workspace class names.

jankeromnes replied:
True -- if we wanted to be really pedantic about this, we could make the dashboard ask server what's the name of the default workspace class 🤔 but yes I'm leaning towards dropping this for now and seeing about it later if needed. 😇

gtsiolis commented on components/dashboard/src/settings/Billing.tsx:
```diff
@@ -13,15 +13,16 @@ import { AttributionId } from "@gitpod/gitpod-protocol/lib/attribution";
     return (
         <PageWithSettingsSubMenu title="Billing" subtitle="Usage-Based Billing.">
             <div>
                 <h3>Billing Account</h3>
```
This could be quite confusing to users, especially now that we've removed the projects under personal accounts and moved the workspaces menu to the left, right?

If it takes more than a few minutes, let's open a follow-up issue for this! 🏓

gtsiolis replied:
To visualize the issue here better, let me post two completely different areas in the dashboard using the same navigation state.

| Billing in User Settings | Projects in Team |
|-|-|
| <img width="1440" alt="Screenshot 2022-10-07 at 3 08 59 PM (2)" src="https://user-images.githubusercontent.com/120486/194550252-d944a9cf-daf0-4321-a328-8906ff355540.png"> | <img width="1440" alt="Screenshot 2022-10-07 at 3 09 02 PM (2)" src="https://user-images.githubusercontent.com/120486/194550292-73d40e4a-0e98-420a-b1e1-1e462663c610.png"> |

jankeromnes replied:
I agree about the confusing part. What would be the correct fix?
- Show whichever team you had selected, but in an "inactive" (i.e. greyed out state)?
- Hide the team selector entirely on user settings pages?

gtsiolis replied:
> Show whichever team you had selected, but in an "inactive" (i.e. greyed out) state?

Yes, which is exactly what we show on every other page under user settings, see example screenshots below.

Later on, we can discuss hiding, replacing, or improving navigation elements, see [relevant discussion](https://gitpod.slack.com/archives/C01KPEPNBD0/p1665147469413099?thread_ts=1665041540.568379&cid=C01KPEPNBD0) (internal).

| Notifications ✔️ | Variables ✔️ | SSH Keys ✔️ | Billing :x: |
|-|-|-|-|
| <img width="1440" alt="Screenshot 2022-10-07 at 3 54 52 PM (2)" src="https://user-images.githubusercontent.com/120486/194558716-e5cd0dc6-a106-4222-8de2-e786b4f642d2.png"> | <img width="1440" alt="Screenshot 2022-10-07 at 3 54 58 PM (2)" src="https://user-images.githubusercontent.com/120486/194558749-8c174348-9ad6-4679-ba78-b62678ec17cf.png"> | <img width="1440" alt="Screenshot 2022-10-07 at 3 54 59 PM (2)" src="https://user-images.githubusercontent.com/120486/194558763-3ee96dbc-14f5-4384-83c2-67fe7fd2fc90.png"> | <img width="1440" alt="Screenshot 2022-10-07 at 3 54 56 PM (2)" src="https://user-images.githubusercontent.com/120486/194558774-f00e2009-1a9b-4f4c-8097-51884a772337.png"> |

jankeromnes commented on components/dashboard/src/components/UsageBasedBillingConfig.tsx:
```diff
@@ -180,55 +200,201 @@ export default function UsageBasedBillingConfig({ subject, attributionId }: Prop
+                            <a className="mt-5 self-end" href={stripePortalUrl}>
                                 <button className="secondary" disabled={!stripePortalUrl}>
-                                    Manage Billing →
+                                    Manage Plan
```
Sure, let's try that 🚀

jankeromnes commented:
_Phew._ Okay I've now addressed most of the comments/suggestions, including fixing the `NaN %` edge case. 🎉

I'm leaving as potential follow-up improvements:
- Improving the handling of usage limits
- Fixing or removing all billing page titles/subtitles
- One thing about a footnote (didn't understand sorry)
- What to do about the two subsequent spinners in team billing

With that, I think we're ready to merge this whenever we want! 🚀 It would be good to deploy this early next week, to support Early Access for Individuals.

/unhold

gtsiolis commented on components/dashboard/src/settings/Billing.tsx:
```diff
@@ -13,15 +13,16 @@ import { AttributionId } from "@gitpod/gitpod-protocol/lib/attribution";
         <PageWithSettingsSubMenu title="Billing" subtitle="Usage-Based Billing.">
             <div>
                 <h3>Billing Account</h3>
                 <BillingAccountSelector />
```
This is mostly about the default state of the cards in there, as well as the text below, which remains confusing and maybe inaccurate, as now you can have billing enabled also for your personal account. Opened https://github.com/gitpod-io/gitpod/issues/13684 to keep track of this.

easyCZ commented on components/dashboard/src/Menu.tsx:
```diff
@@ -97,9 +97,12 @@ export default function Menu() {
     const isMinimalUI = inResource(location.pathname, ["new", "teams/new", "open"]);
     const isWorkspacesUI = inResource(location.pathname, ["workspaces"]);
-    const isAccountUI = inResource(location.pathname, [
+    const isPersonalSettingsUI = inResource(location.pathname, [
```
[random] `inResource` is super misleading. I expected this to check if the pathname contains that resource, but it's actually checking if it starts with those resources.

jankeromnes replied:
Thanks for calling this out @easyCZ! I must admit I only briefly skimmed this code while implementing my fix, but I agree about the name `inResource` being super unclear and unhelpful. Do you have a better suggestion in mind? 🙂

easyCZ replied:
`resourceStartsWith` or `locationStartsWith` comes to mind. Mostly to communicate it's not just a substring match, but a starts with match
